### ISSUE TITLE
Key property excluded from SQL

### ISSUE THREAD

**Issue by flipchart** [artiomchi/FlexLabs.Upsert#6]
When you have an entity with a `Guid` key, that is not set to get its value from the DB, Upsert still excludes it because EF sets the `ValueGenerated` property to `OnAdd`, so the following line:

```
if (prop.ValueGenerated != ValueGenerated.Never) continue;
```

excludes the key and I get a SQL error saying the key column doesn't accept nulls. The key is configured as such:

```
class Entity
{
	public Guid Id { get; set; }
	public string Name { get; set; }
}

class MyContext : DbContext
{
	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		modelBuilder
			.Entity<Entity>()
			.HasKey(x => x.Id);
	}
}
```

No idea why EF specifies it this way, but the model works when inserting with EF normally (i.e. uses the value on the entity, rather than generating it on the DB)

**Comment by artiomchi** [comment]
Sorry it took me a couple weeks to get to this. So quick question, why not use the `[DatabaseGenerated]` attribute on your property? I've just tested this, and it seems to do the trick just fine :)

```csharp
class Entity
{
	[DatabaseGenerated(DatabaseGeneratedOption.None)]
	public Guid Id { get; set; }
	public string Name { get; set; }
}
```

**Comment by artiomchi** [comment]
Since there was no reply to this in months, I'm closing this issue. Feel free to re-open it if the issue persists and the solution above doesn't work for you

**Comment by corstian** [comment]
Hijacking this issue;

Using the 2.0.0 package I'm hitting a similar problem whereas the `Id` property is excluded from the query. See the generated SQL:

``` SQL
MERGE INTO [TSomethings] 
	WITH (HOLDLOCK) AS [T] 
USING ( VALUES (@p0, @p1, @p2, @p3) ) 
	AS [S] ([False], [I], [Moment], [Name]) 
ON [T].[Id] = [S].[Id] 
WHEN NOT MATCHED BY TARGET THEN 
	INSERT ([False], [I], [Moment], [Name]) 
	VALUES ([False], [I], [Moment], [Name]) 
WHEN MATCHED THEN 
	UPDATE SET [False] = [S].[False], [I] = [S].[I], [Moment] = [S].[Moment], [Name] = [S].[Name];
```

The `Id` field is the identity column in this table. The following code is used to generate this statement:

``` C#
var t = new TSomething()
{
    Id = -1,
    False = true
};

await context.TSomethings
    .Upsert(t)
    .RunAsync();
```

Currently unable to push the reproduction project to some repo as GitHub is having troubles. Will try to do so asap.

**Comment by artiomchi** [comment]
Hi Corstian,

Yeah, GitHub's having issues with their storage systems, so I don't even see your comment here.. I did get an e-mail notification though.

Looking at the generated query, the `Id` field is added to the query - it's used in the Merge's `ON` statement. It's not used in the `INSERT` statement because your table has an identity insert clause added to it (by default), which means it's values are generated by SQL server, and you can't manually set it's value (well, not easily, not by default).

If you want to be able to manually pick the values in your column, follow the same approach that I've listed above with the `DatabaseGenerated` attribute. If you do this, you'll have to make sure you have a unique value for EVERY entity you insert into the table.

If you want SQL server to pick unique values to the table, then you won't be able to manually pick the Id values in the INSERT statement - SQL server won't allow it.

**Disclaimer**: You can manually insert entries with custom entered `Id` values by temporarily turning on identity inserts to the table, but you'd want to turn that off as soon as you inserted an entry to the table. That would usually be done manually to add custom entries (like -1) to the table, and wouldn't be used on a regular basis, or by a library using EF, so as such is outside the scope of this project :)

**Comment by corstian** [comment]
I totally follow you there, however, what is the point of an upsert if you cannot determine whether an item with a certain id is already inserted. I imagine using this library by just passing an object instance, and having the ability to keep the identity null or empty for an insert.

(So I'm not trying to manually insert an identity generated Id in the db :))

![image](https://user-images.githubusercontent.com/5021647/47295561-37df4500-d610-11e8-86e4-d2271467a4cc.png)

Above image illustrates what's going wrong with the current generated query.

**Comment by corstian** [comment]
A minimalistic bug reproduction should be available at https://github.com/corstian/efcore-upsert-bug-repro by now.

**Comment by artiomchi** [comment]
oh... umm.... Ok, this looks bad..

Could you show me your table class?

**Comment by artiomchi** [comment]
Oh wait, for a second there I thought it was a big bug :)

Yes, the behaviour is correct. In your case, you have an `Id` column that is database generated. If the column is database generated, you shouldn't be using it when doing a match comparison, since it's value is dynamic. Instead you should be using a different column to try find unique matches.

The reason for that is simple - in your case you're trying to upsert an entity with an `Id` set to -1. If that entry already exists in the database - you can update it, but if it doesn't - the insert will never work, since you can't insert an entity to the database, specifying it's ID.

Think of the following T-SQL.. If you run it multiple times, it will insert a new entry to the database, instead of inserting it only once, since the `Id` will always get a new (positive) value generated

```sql
MERGE INTO [TSomethings] 
	WITH (HOLDLOCK) AS [T] 
USING ( VALUES (@p0, @p1, @p2, @p3, @p4) ) 
	AS [S] ([Id], [False], [I], [Moment], [Name]) 
ON [T].[Id] = [S].[Id] 
WHEN NOT MATCHED BY TARGET THEN 
	INSERT ([False], [I], [Moment], [Name]) 
	VALUES ([False], [I], [Moment], [Name]) 
WHEN MATCHED THEN 
	UPDATE SET [False] = [S].[False], [I] = [S].[I], [Moment] = [S].[Moment], [Name] = [S].[Name];
```

Thinking out loud - I should throw an exception with more details in cases like this :)

**Comment by corstian** [comment]
I'm following your thoughts there. The expected behaviour from my part is that the update and insert functionality of Entity Framework is compacted in a single command with some smartness on the query side. Therefore I expected; 

```C#
var t = new TSomething()
{
    Id = default(int),
    False = true
};
```
To result in an insert, while
```C#
var t = new TSomething()
{
    Id = 1200,
    False = true
};
```
Would result in an update of the entity with id 1200.

I get it that the moment you'd run the first example from this message (which would result in the T-SQL code you just posted) a value is inserted multiple times, but wouldn't that be the choice of the user, for specifying the columns to match on? I still think it'd be a valid use case.

Also, as stated on https://github.com/artiomchi/FlexLabs.Upsert/wiki/Usage; "If the On call is not made, the matching will be done on the primary key on the table.", and therefore I was expecting the specified statement to work.

Also, I really appreciate your quick responses :)

**Comment by corstian** [comment]
For context; this is what currently happens when the Id field is omitted:

![image](https://user-images.githubusercontent.com/5021647/47305594-3d498900-d62a-11e8-97c2-0a099396a764.png)

**Comment by artiomchi** [comment]
Well.. If you created your entity manually and called `context.TSomethings.Add(t)`, then EF would add it, and if you loaded it from the db and changed it - it would call update it :)

EF wouldn't be making that decision based on what you stored in the `Id` property.

Also, when inserting entries, and with a table like yours (with an auto increment `Id`), EF will completely ignore whatever you put in that field when inserting values.

So with a case like this:
```c#
var t = new TSomething()
{
    Id = 1200,
    False = true
};
context.TSomethings.Add(t);
await context.SaveChangesAsync();

Console.WriteLine(t.Id);
```

You'll **not** get 1200 back - instead you'll get the next available `Id` value from the table - the auto increment value. 

If you try to call 
```sql
INSERT INTO [dbo].[TSomethings] ( [Id], [False] ) VALUES ( 1200, 0 )
```

You'll get this exception:
```
Msg 544, Level 16, State 1, Line 1
Cannot insert explicit value for identity column in table 'TSomethings' when IDENTITY_INSERT is set to OFF.
```

THIS is why you can't use the identity column as the `On` key on your table. It's not because of how the Upsert library works - it's because of how Sql works :)

P.S. Thanks!
I usually try to answer quickly whenever possibly, but work can get in the way! 😄

**Comment by corstian** [comment]
Yeah I see! I decided to settle on client side generated guid's as PK anyway so I guess I'm in the clear for the moment! Thanks so far! :smile:

**Comment by artiomchi** [comment]
Sure, no problems! Anything else I can help on this issue, or we can close it?

**Comment by corstian** [comment]
I think it's clear now and we can close it. Maybe it'd be nice to add an exception in the future that's a bit more clear on what's going on ;)

**Comment by artiomchi** [comment]
Yup, absolutely! I'll add it in the next release!
